// ice/src/es5/Ice/Connection.js
"use strict";

function _typeof(obj) { "@babel/helpers - typeof"; return _typeof = "function" == typeof Symbol && "symbol" == typeof Symbol.iterator ? function (obj) { return typeof obj; } : function (obj) { return obj && "function" == typeof Symbol && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }, _typeof(obj); }
function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); Object.defineProperty(subClass, "prototype", { writable: false }); if (superClass) _setPrototypeOf(subClass, superClass); }
function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf ? Object.setPrototypeOf.bind() : function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }
function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = _getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = _getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return _possibleConstructorReturn(this, result); }; }
function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } else if (call !== void 0) { throw new TypeError("Derived constructors may only return object or undefined"); } return _assertThisInitialized(self); }
function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }
function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }
function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf.bind() : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }
function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, _toPropertyKey(descriptor.key), descriptor); } }
function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); Object.defineProperty(Constructor, "prototype", { writable: false }); return Constructor; }
function _toPropertyKey(arg) { var key = _toPrimitive(arg, "string"); return _typeof(key) === "symbol" ? key : String(key); }
function _toPrimitive(input, hint) { if (_typeof(input) !== "object" || input === null) return input; var prim = input[Symbol.toPrimitive]; if (prim !== undefined) { var res = prim.call(input, hint || "default"); if (_typeof(res) !== "object") return res; throw new TypeError("@@toPrimitive must return a primitive value."); } return (hint === "string" ? String : Number)(input); }
function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }
//
// Copyright (c) ZeroC, Inc. All rights reserved.
//
//
// Ice version 3.7.10
//
// <auto-generated>
//
// Generated from file `Connection.ice'
//
// Warning: do not edit this file.
//
// </auto-generated>
//

/* eslint-disable */
/* jshint ignore: start */

/* slice2js browser-bundle-skip */
var _ModuleRegistry = require("../Ice/ModuleRegistry").Ice._ModuleRegistry;
require("../Ice/Object");
require("../Ice/Value");
require("../Ice/ObjectPrx");
require("../Ice/Struct");
require("../Ice/EnumBase");
require("../Ice/Long");
require("../Ice/HashMap");
require("../Ice/HashUtil");
require("../Ice/ArrayUtil");
require("../Ice/StreamHelpers");
require("../Ice/ObjectAdapterF");
require("../Ice/Identity");
require("../Ice/Endpoint");
var Ice = _ModuleRegistry.module("Ice");
var Slice = Ice.Slice;
/* slice2js browser-bundle-skip-end */

/**
 * The batch compression option when flushing queued batch requests.
 *
 **/
Ice.CompressBatch = Slice.defineEnum([['Yes', 0], ['No', 1], ['BasedOnProxy', 2]]);

/**
 * Base class providing access to the connection details.
 *
 **/
Ice.ConnectionInfo = /*#__PURE__*/function () {
  function _class() {
    var underlying = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : null;
    var incoming = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : false;
    var adapterName = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : "";
    var connectionId = arguments.length > 3 && arguments[3] !== undefined ? arguments[3] : "";
    _classCallCheck(this, _class);
    this.underlying = underlying;
    this.incoming = incoming;
    this.adapterName = adapterName;
    this.connectionId = connectionId;
  }
  return _createClass(_class);
}();

/**
 * Specifies the close semantics for Active Connection Management.
 **/
Ice.ACMClose = Slice.defineEnum([['CloseOff', 0], ['CloseOnIdle', 1], ['CloseOnInvocation', 2], ['CloseOnInvocationAndIdle', 3], ['CloseOnIdleForceful', 4]]);

/**
 * Specifies the heartbeat semantics for Active Connection Management.
 **/
Ice.ACMHeartbeat = Slice.defineEnum([['HeartbeatOff', 0], ['HeartbeatOnDispatch', 1], ['HeartbeatOnIdle', 2], ['HeartbeatAlways', 3]]);

/**
 * A collection of Active Connection Management configuration settings.
 **/
Ice.ACM = /*#__PURE__*/function () {
  function _class2() {
    var timeout = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : 0;
    var close = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : Ice.ACMClose.CloseOff;
    var heartbeat = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : Ice.ACMHeartbeat.HeartbeatOff;
    _classCallCheck(this, _class2);
    this.timeout = timeout;
    this.close = close;
    this.heartbeat = heartbeat;
  }
  return _createClass(_class2);
}();
Slice.defineStruct(Ice.ACM, true, true);

/**
 * Determines the behavior when manually closing a connection.
 **/
Ice.ConnectionClose = Slice.defineEnum([['Forcefully', 0], ['Gracefully', 1], ['GracefullyWithWait', 2]]);

/**
 * Provides access to the connection details of an IP connection
 *
 **/
Ice.IPConnectionInfo = /*#__PURE__*/function (_Ice$ConnectionInfo) {
  _inherits(_class3, _Ice$ConnectionInfo);
  var _super = _createSuper(_class3);
  function _class3(underlying, incoming, adapterName, connectionId) {
    var _this;
    var localAddress = arguments.length > 4 && arguments[4] !== undefined ? arguments[4] : "";
    var localPort = arguments.length > 5 && arguments[5] !== undefined ? arguments[5] : -1;
    var remoteAddress = arguments.length > 6 && arguments[6] !== undefined ? arguments[6] : "";
    var remotePort = arguments.length > 7 && arguments[7] !== undefined ? arguments[7] : -1;
    _classCallCheck(this, _class3);
    _this = _super.call(this, underlying, incoming, adapterName, connectionId);
    _this.localAddress = localAddress;
    _this.localPort = localPort;
    _this.remoteAddress = remoteAddress;
    _this.remotePort = remotePort;
    return _this;
  }
  return _createClass(_class3);
}(Ice.ConnectionInfo);

/**
 * Provides access to the connection details of a TCP connection
 *
 **/
Ice.TCPConnectionInfo = /*#__PURE__*/function (_Ice$IPConnectionInfo) {
  _inherits(_class4, _Ice$IPConnectionInfo);
  var _super2 = _createSuper(_class4);
  function _class4(underlying, incoming, adapterName, connectionId, localAddress, localPort, remoteAddress, remotePort) {
    var _this2;
    var rcvSize = arguments.length > 8 && arguments[8] !== undefined ? arguments[8] : 0;
    var sndSize = arguments.length > 9 && arguments[9] !== undefined ? arguments[9] : 0;
    _classCallCheck(this, _class4);
    _this2 = _super2.call(this, underlying, incoming, adapterName, connectionId, localAddress, localPort, remoteAddress, remotePort);
    _this2.rcvSize = rcvSize;
    _this2.sndSize = sndSize;
    return _this2;
  }
  return _createClass(_class4);
}(Ice.IPConnectionInfo);

/**
 * Provides access to the connection details of a UDP connection
 *
 **/
Ice.UDPConnectionInfo = /*#__PURE__*/function (_Ice$IPConnectionInfo2) {
  _inherits(_class5, _Ice$IPConnectionInfo2);
  var _super3 = _createSuper(_class5);
  function _class5(underlying, incoming, adapterName, connectionId, localAddress, localPort, remoteAddress, remotePort) {
    var _this3;
    var mcastAddress = arguments.length > 8 && arguments[8] !== undefined ? arguments[8] : "";
    var mcastPort = arguments.length > 9 && arguments[9] !== undefined ? arguments[9] : -1;
    var rcvSize = arguments.length > 10 && arguments[10] !== undefined ? arguments[10] : 0;
    var sndSize = arguments.length > 11 && arguments[11] !== undefined ? arguments[11] : 0;
    _classCallCheck(this, _class5);
    _this3 = _super3.call(this, underlying, incoming, adapterName, connectionId, localAddress, localPort, remoteAddress, remotePort);
    _this3.mcastAddress = mcastAddress;
    _this3.mcastPort = mcastPort;
    _this3.rcvSize = rcvSize;
    _this3.sndSize = sndSize;
    return _this3;
  }
  return _createClass(_class5);
}(Ice.IPConnectionInfo);
Slice.defineDictionary(Ice, "HeaderDict", "HeaderDictHelper", "Ice.StringHelper", "Ice.StringHelper", false, undefined, undefined);

/**
 * Provides access to the connection details of a WebSocket connection
 *
 **/
Ice.WSConnectionInfo = /*#__PURE__*/function (_Ice$ConnectionInfo2) {
  _inherits(_class6, _Ice$ConnectionInfo2);
  var _super4 = _createSuper(_class6);
  function _class6(underlying, incoming, adapterName, connectionId) {
    var _this4;
    var headers = arguments.length > 4 && arguments[4] !== undefined ? arguments[4] : null;
    _classCallCheck(this, _class6);
    _this4 = _super4.call(this, underlying, incoming, adapterName, connectionId);
    _this4.headers = headers;
    return _this4;
  }
  return _createClass(_class6);
}(Ice.ConnectionInfo);

/* slice2js browser-bundle-skip */
exports.Ice = Ice;
/* slice2js browser-bundle-skip-end */
